### ISSUE TITLE
"profile:me" or "jwt" OAuth scope

### ISSUE THREAD

**Issue by etodd** [itchio/itch.io#705]
My game needs a way to authenticate users if it is launched some way other than the itch app. I'd like to use the same OAuth login flow that butler does, but don't want to save a token that has access to all the wharf functionality. It only needs access to a single API call that generates a JWT token for use with /api/1/jwt/me.

**Comment by leafo** [comment]
The way the butler oauth flow works right now is by using a browser redirect back to localhost to complete passing the token. This means there's a temporary webserver running on the computer. Does this sound like something you would do, or would you want the user to have to copy and paste a token into your game?

**Comment by etodd** [comment]
Yep, temporary webserver is totally fine!

**Comment by fasterthanlime** [comment]
Woop my bad @leafo, Evan & I had the exact same exchange by DM, I should've summed it up here :)

**Comment by etodd** [comment]
Any word on this? I've got a little mongoose server set up ready to go.

**Comment by fasterthanlime** [comment]
Hey Evan, I'll give it a shot tomorrow!

**Comment by fasterthanlime** [comment]
I've been reviewing our current OAuth2 flow for butler and found three deviations from the spec (thanks very much, past me).

---

1. butler hardcodes its `client_id` to `butler` when it should really be something like a UUID - but that's actually good - it allows us to detect "old butler versions" and serve them the slightly-deviated oauth version
2. the callback URL we redirect to does `whatever/oauth/callback#XXX` instead of `whatever/oauth/callback#token=XXX` - I'll fix that, will need to keep the old way for client_id=butler for a bit
3. we don't force pre-registering redirect URLs, even though the oauth 2.0 spec is what makes the implicit flow "trustworthy"

The problem with 3. is that forcing you to pick a port in advance might not work - what if another application is already listening on this port on the user's computer?

---

I've been looking around a bit, and there's a google extension that kinda bypasses the need for an http server, for Windows:

```
**urn:ietf:wg:oauth:2.0:oob:auto**

This is identical to urn:ietf:wg:oauth:2.0:oob, but the text in the confirmation page that the OAuth 2.0 server generates won't instruct the user to copy the authorization code, but instead will simply ask the user to close the window.

This is useful when your application reads the title of the HTML page (by checking window titles on the desktop, for example) to obtain the authorization code, but can't close the page on its own.
```

see https://developers.google.com/api-client-library/python/auth/installed-app

Interestingly, the itch.io sandbox would prevent you from doing that (reading other windows' titles) - but that's irrelevant since the use-case here is to run the game outside the itch.io app :)

**Comment by fasterthanlime** [comment]
So, imho the cleanest solution would be for you to provide a redirect URL that's served by your game server, not by the game itself. It could go something like this:

  - GameClient contacts GameServer, doesn't have credentials
  - GameServer says "ok your login request has id <UUID>, you can poll this endpoint for credentials, please open url <URL> for your user to log in"
  - GameClient opens <URL> in user's browser
  - User approves access, is redirected to callback URL on GameServer
  - GameServer serves a script that reads the token from the URL hash & POSTs it back to GameServer

Here's how butler does that step:

https://github.com/itchio/butler/blob/6f54272227df8dfafa75b99f8f7d0624c7a1b3ff/authenticate.go#L75-L88

  - GameClient, who has been polling GameServer every two seconds or so, finally gets its credentials and can use that!

Does that sound reliable to you @etodd ?

On one hand, you have to do a bunch of work on your server, but on the other hand, you don't even need an http server in the game (fewer crashes & you won't need to explain to users that they need to click 'Yes' on that Windows firewall window that pops up *behind* your game), the user doesn't need to copy-paste anything, and it'll work on Windows/macOS/Linux just the same

**Comment by fasterthanlime** [comment]
I feel like I need to explain a bit more why pre-registering redirect URLs for the implicit flow is important:

The `client_id` (that you need to include as a query parameter to the /login URL) is not a secret - anyone trying to log in via your game is going to be able to see it.

If we accept just any redirect URI (or even a hardcoded `localhost` one tbh), then anybody would be able to do the login flow with it and do requests that will be associated to your game. (There's no client secret in the implicit flow - shipping a secret in a binary is pointless anyway, since it's easily reversed).

**Comment by fasterthanlime** [comment]
Reading Google docs a bit more, it looks like they actually recommend the localhost http server approach over the others now, see https://developers.google.com/identity/protocols/OAuth2InstalledApp#request-parameter-redirect_uri

and:

> **Important:** The [custom URI scheme](https://developers.google.com/identity/protocols/OAuth2InstalledApp#redirect-uri_custom-scheme) and [loopback IP address](https://developers.google.com/identity/protocols/OAuth2InstalledApp#redirect-uri_loopback) options now provide more reliable, secure, and user-friendly ways to obtain user authorization. This option may be discontinued in the future and should only be used if the choices above are not viable.

**Comment by etodd** [comment]
I'm fine redirecting to the GameServer or GameClient, either will work. The GameClient localhost method will be easier. My users already have to click through the firewall popup anyway.

I understand the need for pre-registered URLs. Does that mean itch.io will need a whole CRUD interface for those? Yikes.

**Comment by fasterthanlime** [comment]
> I understand the need for pre-registered URLs. Does that mean itch.io will need a whole CRUD interface for those? Yikes.

Yes it does! I'm doing it right this moment :)

**Comment by fasterthanlime** [comment]
Most of the feature is done, it's pending review, here's how it looks from my side:

---

  * New app dialog:

![oauth1](https://user-images.githubusercontent.com/7998310/30978362-a178ee12-a47a-11e7-882a-17be3e87628b.png)

---

  * App list:

![oauth2](https://user-images.githubusercontent.com/7998310/30978396-ba4a4b0c-a47a-11e7-826f-949abf2723f2.png)

---

  * App view/edit:

![oauth3](https://user-images.githubusercontent.com/7998310/30978417-c6af0afe-a47a-11e7-8679-6745e87bdb40.png)

---

  * Login page:

![oauth4](https://user-images.githubusercontent.com/7998310/30978429-cd343b74-a47a-11e7-806b-df977e235416.png)

---

  * How it shows up in user settings for user of your app:

![oauth5](https://user-images.githubusercontent.com/7998310/30978444-d7164e48-a47a-11e7-9d78-f4f9622a8040.png)

---

Docs have also been improved - they're still very much lacking, but should cover all you've requested here:

![oauth6](https://user-images.githubusercontent.com/7998310/30978527-1587f208-a47b-11e7-975d-deacc982526c.png)

---

In server-side API, endpoints now show their scope:

![oauth7](https://user-images.githubusercontent.com/7998310/30978530-1a2d3ca0-a47b-11e7-9b13-d9ac3468c122.png)

**Comment by etodd** [comment]
Wow, this is fantastic! This is a huge amount of UI and docs! Thank you so much!

Now if we could juuuust get a tiny endpoint that generates a JWT, that will complete the whole auth process! :smile: :smile:

**Comment by fasterthanlime** [comment]
> Now if we could juuuust get a tiny endpoint that generates a JWT, that will complete the whole auth process! 😄 😄

I've actually documented this as well - API keys & JWT tokens both give you access to the same endpoints (assuming you've got equivalent scopes for both) - you just use them slightly differently.

The itch.io app has an API key (which it gets when you log into the app) and it uses an endpoint to generate more narrowly-scoped JWT token, which it passes to your game (when you ask via manifest).

I can make that endpoint available to OAuth apps if you want the game to deal only with JWT tokens :) But again, both are valid credentials.

**Comment by fasterthanlime** [comment]
@etodd here's the relevant bit of the (yet-unpublished) docs:

---

Authentication
--------------

All API endpoints require authentication.

There are two types of credentials you can use:

-   API keys are long-lasting credentials that can be revoked by users
-   JWT tokens are short-lived, expiring credentials

JWT tokens expire but they're easier to obtain

### API keys

You can generate API keys for your own account from the [API keys](http://localhost.com:8080/api-keys) page. (You'll need an [itch.io account](http://localhost.com:8080/developers) if you don't already have one.)

If you want to make requests on behalf of other itch.io users, you can register an [OAuth application](http://localhost.com:8080/docs/api/oauth) and have them grant permissions to your app.

To use an API key to authenticate an API request, simply replace `KEY` in the following API endpoints with your API key.

### JWT tokens

JWT tokens can be obtained by specifying

To use a JWT token to authenticate an API request, replace `KEY` in the following endpoints with `jwt`, and add the following HTTP header:

```
Authorization: YOUR_JWT_TOKEN
```

Scopes
------

Credentials (whether they're API keys or JWT tokens) can have a limited scope, which means they'll give access to some endpoints but not others.

In the API reference below, endpoints are listed with: HTTP method (in blue), scope (in green), and template URL.

Having access to a scope gives access to all its subscopes. For example, `profile` gives access to `profile:me`, but `game:view:purchases` does not give access to `game:view`.

> API keys that you generate for yourself are unscoped -- they have access to all endpoints.

**Comment by etodd** [comment]
Ah, perfect! Yeah, I think it would be good to allow OAuth apps to create a more narrow-scoped JWT, just for that extra tiny bit of security. And that way, authentication always works the same on the server-side of the game, instead of having to differentiate between API keys and JWTs.

**Comment by etodd** [comment]
Any news on this getting released? I'm itching to launch! :)

**Comment by leafo** [comment]
Sorry about the delay, I was out of town and didn't get a chance to review it. It should be going out today

**Comment by leafo** [comment]
Alright, it's deployed now: https://itch.io/docs/api/oauth

tell us if you have any issues

**Comment by fasterthanlime** [comment]
@etodd a few notes!

You'll have to support both jwt tokens (from app) and API keys, but I made them more consistent.

All you need is to remember if it was a jwt token and an API key and then you do:

```
GET /api/1/jwt/...
Authorization: Bearer :value
```

or:

```
GET /api/1/key/...
Authorization: Bearer :value
```

Same header - it's just the third path component that changes.

I wanted to expose subkey for you but due to how the app v23 behaves it wasn't doable without opening a potential security hole!

**Comment by etodd** [comment]
I appreciate the API consistency! :) Got everything working just fine. You're all awesome. Thank you so much.

**Comment by quyse** [comment]
This is great!

I have a suggestion: maybe itch should not ask user every time for permission, and starting from the second time just silently redirect to callback URL, as the token is generated only once per oauth app anyway?

Also would be nice to have a way to change icon displayed on the authorization page: currently it's the author's avatar, but I would like to put there my project's icon. Maybe even allow to link registered OAuth application with a project on itch.io, so it can display the project's icon.

Thanks!
